Content.Server/Chemistry/EntitySystems/SolutionInjectOnEventSystem.cs

```diff
@@ -0,0 +1,147 @@
+        SubscribeLocalEvent<MeleeChemicalInjectorComponent, MeleeHitEvent>(HandleMeleeHit);
+    }
+
+    private void HandleProjectileHit(Entity<SolutionInjectOnProjectileHitComponent> entity, ref ProjectileHitEvent args)
+    {
+        DoInjection((entity.Owner, entity.Comp), args.Target, args.Shooter);
+    }
+
+    private void HandleEmbed(Entity<SolutionInjectOnEmbedComponent> entity, ref EmbedEvent args)
+    {
+        DoInjection((entity.Owner, entity.Comp), args.Embedded, args.Shooter);
+    }
+
+    private void HandleMeleeHit(Entity<MeleeChemicalInjectorComponent> entity, ref MeleeHitEvent args)
+    {
+        // MeleeHitEvent is weird, so we have to filter to make sure we actually
+        // hit something and aren't just examining the weapon.
+        if (args.IsHit)
+            TryInjectTargets((entity.Owner, entity.Comp), args.HitEntities, args.User);
+    }
+
+    private void DoInjection(Entity<BaseSolutionInjectOnEventComponent> injectorEntity, EntityUid target, EntityUid? source = null)
+    {
+        TryInjectTargets(injectorEntity, [target], source);
+    }
+
+    /// <summary>
+    /// Filters <paramref name="targets"/> for valid targets and tries to inject a portion of <see cref="BaseSolutionInjectOnEventComponent.Solution"/> into
+    /// each valid target's bloodstream.
+    /// </summary>
+    /// <remarks>
+    /// Targets are invalid if any of the following are true:
+    /// <list type="bullet">
+    ///     <item>The target does not have a bloodstream.</item>
+    ///     <item><see cref="BaseSolutionInjectOnEventComponent.PierceArmor"/> is false and the target is wearing a hardsuit.</item>
+    ///     <item><see cref="BaseSolutionInjectOnEventComponent.BlockSlots"/> is not NONE and the target has an item equipped in any of the specified slots.</item>
+    /// </list>
+    /// </remarks>
+    /// <returns>true if at least one target was successfully injected, otherwise false</returns>
+    private bool TryInjectTargets(Entity<BaseSolutionInjectOnEventComponent> injector, IReadOnlyList<EntityUid> targets, EntityUid? source = null)
+    {
+        // Make sure we have at least one target
+        if (targets.Count == 0)
+            return false;
+
+        // Get the solution to inject
+        if (!_solutionContainer.TryGetSolution(injector.Owner, injector.Comp.Solution, out var injectorSolution))
+            return false;
+
+        // Build a list of bloodstreams to inject into
+        var targetBloodstreams = new List<Entity<BloodstreamComponent>>();
+        foreach (var target in targets)
+        {
+            if (Deleted(target))
+                continue;
+
+            // Yuck, this is way to hardcodey for my tastes
+            // TODO blocking injection with a hardsuit should probably done with a cancellable event or something
+            if (!injector.Comp.PierceArmor && _inventory.TryGetSlotEntity(target, "outerClothing", out var suit) && _tag.HasTag(suit.Value, "Hardsuit"))
```

Comment:
Like the comment says, this is kinda gross. I'm just carrying it over directly from the old melee injection system though; a future PR can replace it with something better.